File: egui_states/logging.py
from collections.abc import Callable
from enum import Enum

from egui_states._core import StateServerCore
from egui_states.signals import SignalsManager


class LogLevel(Enum):
    """Logging levels."""

    Debug = 0
    Info = 1
    Warning = 2
    Error = 3


class LoggingSignal:
    """Logging signal for processing log messages from the state server."""

    def __init__(self, signals_manager: SignalsManager, server: StateServerCore) -> None:
        """Initialize the LoggingSignal."""
        self._loggers: dict[int, list[Callable[[str], None]]] = {0: [], 1: [], 2: [], 3: []}
        logging_id = server.signal_get_logging_id()
        signals_manager.add_callback(logging_id, self._callback)
        server.signal_set_to_multi(logging_id)

    def _callback(self, message: tuple[int, str]) -> None:
        level = message[0]
        if level == LogLevel.Debug.value:
            for logger in self._loggers[0]:
                logger(message[1])
        elif level == LogLevel.Info.value:
            for logger in self._loggers[1]:
                logger(message[1])
        elif level == LogLevel.Warning.value:
            for logger in self._loggers[2]:
                logger(message[1])
        elif level == LogLevel.Error.value:
            for logger in self._loggers[3]:
                logger(message[1])

    def add_logger(self, level: LogLevel, logger: Callable[[str], None]):
        """Add logger for a specific level.

        Args:
            level(Level): The logging level.
            logger(Callable[[str], None]): The logger to add.
        """
        self._loggers[level.value].append(logger)

    def remove_logger(self, level: LogLevel, logger: Callable[[str], None]):
        """Remove logger for a specific level.

        Args:
            level(Level): The logging level.
            logger(Callable[[str], None]): The logger to remove.
        """
        if logger in self._loggers[level.value]:
            self._loggers[level.value].remove(logger)

    def remove_all_loggers(self, level: LogLevel):
        """Remove all loggers for a specific level.

        Args:
            level(Level): The logging level.
        """
        self._loggers[level.value].clear()

File: crates/egui-states/src/server/logging.rs
use crate::server_core::signals::LOGGING_ID;

use super::callbacks::CallbackHandle;
use super::state_server::{StateServer, deserialize_bytes};

#[derive(Clone, Copy, Debug, Eq, PartialEq)]
/// Severity attached to a log message sent by the client.
pub enum LogLevel {
    /// Diagnostic information useful while debugging.
    Debug,
    /// Routine informational output.
    Info,
    /// A potentially problematic condition.
    Warning,
    /// An error reported by the client.
    Error,
}

impl LogLevel {
    fn from_id(id: u8) -> Option<Self> {
        match id {
            0 => Some(Self::Debug),
            1 => Some(Self::Info),
            2 => Some(Self::Warning),
            3 => Some(Self::Error),
            _ => None,
        }
    }
}

#[derive(Clone)]
/// Dispatcher for log messages emitted by the connected client.
pub struct LoggingSignal {
    server: StateServer,
}

impl LoggingSignal {
    /// Creates a logging dispatcher registered with `server`.
    pub fn new(server: &StateServer) -> Self {
        server.set_signal_to_queue(LOGGING_ID);
        Self {
            server: server.clone(),
        }
    }

    /// Registers a callback for one severity level.
    ///
    /// The returned handle must be retained to keep the callback connected.
    pub fn add_logger(
        &self,
        level: LogLevel,
        logger: impl Fn(String) + Send + Sync + 'static,
    ) -> CallbackHandle {
        self.server.add_raw_callback(LOGGING_ID, move |data, _| {
            let (raw_level, message) = deserialize_bytes::<(u8, String)>(&data)?;
            if LogLevel::from_id(raw_level) == Some(level) {
                logger(message);
            }
            Ok(())
        })
    }
}
